kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandEncryptionManager.java

```diff
@@ -0,0 +1,276 @@
+    */
+    private final EncryptionVersion encryptionVersion;
+    private final DekManager<K, E> dekManager;
+    private final AsyncLoadingCache<CacheKey<K>, Dek<E>> dekCache;
+
+    private final int recordBufferInitialBytes;
+    private final int recordBufferMaxBytes;
+
+    public InBandEncryptionManager(@NonNull DekManager<K, E> dekManager,
+                                   int recordBufferInitialBytes,
+                                   int recordBufferMaxBytes,
+                                   @Nullable Executor dekCacheExecutor,
+                                   int dekCacheMaxItems) {
+        this.encryptionVersion = EncryptionVersion.V1; // TODO read from config
+        this.dekManager = Objects.requireNonNull(dekManager);
+        if (recordBufferInitialBytes <= 0) {
+            throw new IllegalArgumentException();
+        }
+        this.recordBufferInitialBytes = recordBufferInitialBytes;
+        if (recordBufferMaxBytes <= 0) {
+            throw new IllegalArgumentException();
+        }
+        this.recordBufferMaxBytes = recordBufferMaxBytes;
+        Caffeine<Object, Object> cache = Caffeine.newBuilder();
+        if (dekCacheMaxItems != NO_MAX_CACHE_SIZE) {
+            cache = cache.maximumSize(dekCacheMaxItems);
+        }
+        if (dekCacheExecutor != null) {
+            cache = cache.executor(dekCacheExecutor);
+        }
+        this.dekCache = cache
+                .removalListener(this::afterCacheEviction)
+                .buildAsync(this::requestGenerateDek);
+    }
+
+    /** Invoked by Caffeine when a DEK needs to be loaded */
+    private CompletableFuture<Dek<E>> requestGenerateDek(@NonNull CacheKey<K> cacheKey,
+                                                         @NonNull Executor executor) {
+        return dekManager.generateDek(cacheKey.kek(), cacheKey.cipherSpec())
+                .thenApply(dek -> {
+                    if (LOGGER.isTraceEnabled()) {
+                        LOGGER.trace("Adding DEK to cache: {}", dek);
+                    }
+                    dek.destroyForDecrypt();
+                    return dek;
+                })
+                .toCompletableFuture();
+    }
+
+    /** Invoked by Caffeine after a DEK is evicted from the cache. */
+    private void afterCacheEviction(@Nullable CacheKey<K> cacheKey,
+                                    @Nullable Dek<E> dek,
+                                    RemovalCause removalCause) {
+        if (dek != null) {
+            dek.destroyForEncrypt();
+            if (LOGGER.isTraceEnabled()) {
+                LOGGER.trace("Attempted to destroy DEK: {}", dek);
+            }
+        }
+    }
+
+    // @VisibleForTesting
+    CompletionStage<Dek<E>> currentDek(@NonNull EncryptionScheme<K> encryptionScheme) {
+        // todo should we add some scheduled timeout as well? or should we rely on the KMS to timeout appropriately.
+        return dekCache.get(cacheKey(encryptionScheme));
```

Comment:
We should switch back to the eventloop after getting from the cache if it is not immediately completed. The cache load could be being completed asynchronously by a KMS operation on another thread. We could do it after the resolveAll. (I'll go check what thread it is completing on, we do configure up an executor for caffeine but I'd guess it's not used for completing the future from get).

Reply:
I chucked logs into the completions and ran a produce session where I pushed 3 messages through one by one on the same connection. Then consumed them all once.

```
encrypt cache completion: ForkJoinPool.commonPool-worker-2
encrypt cache completion: epollEventLoopGroup-5-4
encrypt cache completion: epollEventLoopGroup-5-4
decrypt cache completion: ForkJoinPool.commonPool-worker-2
```
A bit interesting that we're on FJP, I expected to see the HTTP client pool there, maybe caffeine does change executor, I know the FJP is used by caffeine by default.

Reply:
I thought the internal Java HTTP client ran on the FJP anyway rather than a dedicated pool.

Reply:
looks like HttpClientImpl makes it's own cached thread pool if not specified:
```
 Executor ex = builder.executor;
        if (ex == null) {
            ex = Executors.newCachedThreadPool(new DefaultThreadFactory(id));
            isDefaultExecutor = true;
        } else {
            isDefaultExecutor = false;
        }
```

with naming like `namePrefix = "HttpClient-" + clientID + "-Worker-"`